### ISSUE TITLE
Job gets restarted even though it has backoffLimit: 0

### ISSUE THREAD

**Issue by graywolf-at-work** [kubernetes/kubernetes#102959]
#### What happened:

Job was executed, however when pod failed it was started up again.

#### What you expected to happen:

Job with backoffLimit: 0 should not be started second time.

#### How to reproduce it (as minimally and precisely as possible):

Create Job with (at least) following in the following in the spec section:

```
  backoffLimit: 0
  completions: 1
  parallelism: 1
```

It will sometimes create two pods.

#### Anything else we need to know?:

I believe this was caused by the node running the original (first) pod being
auto-killed during cluster downscale:

```
597:2021-06-15T10:11:42.000Z I0615 10:11:42.563881       1 scheduler.go:609] "Successfully bound pod to node" pod="ci-diff/ci-python-1623751902-7bd425d3-h86g8" node="ip-10-1-99-236.eu-west-2.compute.internal" evaluatedNodes=11 feasibleNodes=6
612:2021-06-15T10:13:05.000Z I0615 10:13:05.589809       1 scheduler.go:609] "Successfully bound pod to node" pod="ci-diff/ci-python-1623751902-7bd425d3-97n6z" node="ip-10-1-114-197.eu-west-2.compute.internal" evaluatedNodes=11 feasibleNodes=4
```

```
630:2021-06-15T10:14:12.000Z I0615 10:14:12.325110       1 node_tree.go:100] Removed node "ip-10-1-99-236.eu-west-2.compute.internal" in group "eu-west-2:\x00:eu-west-2a" from NodeTree
```

Naturally I am not sure about this, but seems too suspicious to be a
coincidence.

#### Environment:
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.10", GitCommit:"98d5dc5d36d34a7ee13368a7893dcb400ec4e566", GitTreeState:"clean", BuildDate:"2021-04-15T03:28:42Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.8-eks-96780e", GitCommit:"96780e1b30acbf0a52c38b6030d7853e575bcdf3", GitTreeState:"clean", BuildDate:"2021-03-10T21:32:29Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:

AWS

- OS (e.g: `cat /etc/os-release`):

`Amazon Linux 2`

- Kernel (e.g. `uname -a`):

```
Linux XXXXXXXXXXXXXXXXXXX 5.4.105-48.177.amzn2.x86_64 #1 SMP Tue Mar 16 04:56:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

- Install tools:

N/A

- Network plugin and version (if this is a network-related bug):

N/A

- Others:

N/A

**Comment by ahg-g** [comment]
This is probably related to https://github.com/kubernetes/kubernetes/issues/28486, which we will be fixing in this cycle (https://github.com/kubernetes/kubernetes/pull/98238). When a node is downscaled, the pod becomes orphaned and gets completely removed, and the job controller doesn't count it as a failure.

When this happens, what do you observe in the [Failed](https://github.com/kubernetes/kubernetes/blob/33c4383985c9dd86bbdff01cfcb9e582accb577d/pkg/apis/batch/types.go#L245) field in JobStatus? If still 0, then #28486 is likely the issue.

**Comment by graywolf-at-work** [comment]
I've put in place some countermeasures for this to not happen again, so I've
based my analysis just on the logs I was able to get from the cluster, but it
indeed looks like what you described is accurate. Near the end I see

```
    "status": {
      "conditions": [
        {
          "type": "Failed",
          "status": "True",
          "lastProbeTime": "2021-06-15T10:13:23Z",
          "lastTransitionTime": "2021-06-15T10:13:23Z",
          "reason": "BackoffLimitExceeded",
          "message": "Job has reached the specified backoff limit"
        }
      ],
      "startTime": "2021-06-15T10:11:42Z",
      "failed": 1
    }
  },
  "requestReceivedTimestamp": "2021-06-15T10:13:23.785105Z",
  "stageTimestamp": "2021-06-15T10:13:23.793924Z",
```

But this is after 2 pods were done finished (based on logs from loki), so looks
like it really does not know about the first one. I'm not able to find more in
the logs.

Looking forward to 22 for this being fixed.

**Comment by alculquicondor** [comment]
Note that 1.22 has this feature as alpha, meaning that it's disabled by default. Please try it out if you have the chance and let us know if you encounter any issues.

**Comment by graywolf-at-work** [comment]
Since we are using EKS (AWS service), we are sadly not able to enable alpha features.

**Comment by graywolf-at-work** [comment]
.

**Comment by graywolf-at-work** [comment]
Maybe there should be a bot that comments with the `/remove-lifecycle stale`...

**Comment by alculquicondor** [comment]
I don't think there's anything actionable here. You could try JobTrackingWithFinalizers in 1.23, if the feature is not disabled first https://github.com/kubernetes/kubernetes/pull/109491

**Comment by alculquicondor** [comment]
Not really, the point is that is it reminds us to pay attention to the issue.

However, as I already said, I don't think there is anything to do for this.

This is already fixed in 1.25+, or in 1.23+ if you enable the feature gate JobTrackingWithFinalizers.

/close
